src/Arcus.WebApi.Logging/RequestTrackingMiddleware.cs

```diff
@@ -54,35 +59,73 @@ public async Task Invoke(HttpContext httpContext)
                 httpContext.Request.EnableBuffering();
                 requestBody = await GetRequestBodyAsync(httpContext);
             }
-
-            try
+            
+            Stream originalResponseBodyStream = null;
+            using (Stream temporaryResponseBodyStream = DetermineResponseBodyBuffer())
             {
-                await _next(httpContext);
+                if (_options.IncludeResponseBody)
+                {
+                    originalResponseBodyStream = httpContext.Response.Body;
+                    httpContext.Response.Body = temporaryResponseBodyStream;
+                }
+            
+                try
+                {
+                    await _next(httpContext);
+                }
+                finally
+                {
+                    string responseBody = await GetResponseBodyAsync(httpContext);
+                    
+                    stopwatch.Stop();
+                    TrackRequest(requestBody, responseBody, httpContext, stopwatch.Elapsed);
+
+                    if (_options.IncludeResponseBody)
+                    {
+                        temporaryResponseBodyStream.Position = 0;
+                        
+                        // Copy back the seekable/temporary response body stream to the original response body stream,
+                        // for the remaining middleware components that comes after this one.
+                        await temporaryResponseBodyStream.CopyToAsync(originalResponseBodyStream); 
+                    }
+                }
             }
-            finally
+        }
+        
+        private Stream DetermineResponseBodyBuffer()
+        {
+            if (_options.IncludeResponseBody)
             {
-                stopwatch.Stop();
-                TrackRequest(requestBody, httpContext, stopwatch.Elapsed);
+                var responseBodyBuffer = new MemoryStream();
```

Comment:
Something like a virtual stream that overloads to disk or something may be better than just a memory stream.
Or maybe something like this: https://www.nuget.org/packages/Microsoft.IO.RecyclableMemoryStream/